File: examples/tool-use.ts
/**
 * Capture tool use events during a query.
 *
 * When forge runs in verbose mode, tool executions appear as status lines
 * in stdout. The SDK parses these into `tool_use` messages with the
 * tool name and arguments.
 *
 * Run: bun run examples/tool-use.ts
 */
import { query } from "../src";

const toolCalls: Array<{ name: string; args: Record<string, unknown> }> = [];

for await (const message of query({
  prompt: "List the files in the current directory, then read package.json.",
  options: {
    cwd: import.meta.dir,
  },
})) {
  switch (message.type) {
    case "system":
      console.log(`[session] ${message.session_id}`);
      break;
    case "assistant":
      process.stdout.write(message.content);
      break;
    case "tool_use":
      toolCalls.push({ name: message.name, args: message.arguments });
      console.log(`\n[tool_use] ${message.name}(${JSON.stringify(message.arguments)})`);
      break;
    case "result":
      console.log(`\n[result] ${message.result.slice(0, 200)}...`);
      break;
    case "error":
      console.error(`[error] ${message.error}`);
      break;
  }
}

console.log(`\nTotal tool calls: ${toolCalls.length}`);
for (const call of toolCalls) {
  console.log(`  - ${call.name}`);
}

File: sdks/python/examples/tool_use.py
"""
Capture tool use events during a query.

When forge runs in verbose mode, tool executions appear as status lines
in stdout. The SDK parses these into `tool_use` messages with the
tool name and arguments.

Run: uv run python examples/tool_use.py
"""
import asyncio
import os
import sys

from forgecode import query


async def main() -> None:
    tool_calls: list[dict[str, object]] = []

    async for message in query(
        "List the files in the current directory, then read package.json.",
        options={"cwd": os.getcwd()},
    ):
        match message.type:
            case "system":
                print(f"[session] {message.session_id}")
            case "assistant":
                sys.stdout.write(message.content)
            case "tool_use":
                tool_calls.append({"name": message.name, "args": message.arguments})
                print(f"\n[tool_use] {message.name}({message.arguments})")
            case "result":
                print(f"\n[result] {message.result[:200]}...")
            case "error":
                print(f"[error] {message.error}", file=sys.stderr)

    print(f"\nTotal tool calls: {len(tool_calls)}")
    for call in tool_calls:
        print(f"  - {call['name']}")


if __name__ == "__main__":
    asyncio.run(main())